### ISSUE TITLE
Save load during CH4 causes crash on real hardware

### ISSUE THREAD

**Issue by Turom** [wagrenier/Zero2Undub#6]
The crash happens during the loading screen after selecting a save during chapter 4.

On real hardware the game freezes. On PCSX2 the log displays errors but the game loads anyway.

When loading the same save file with unpatched version there is no crash and no errors both on real hardware and PCSX2. When attempting to load the save with unpatched version and create a new save then go back to patched version, the error persists.

Test environment:
PCSX2 version 1.6.0
BIOS  SCPH-90001 NA 230-080220
[PCSX2 Memory Card](https://hosting.thegenesiaproject.com/turom/Mcd001.ps2)
[PSU save file](https://hosting.thegenesiaproject.com/turom/FATAL%20FRAME2GameData.psu)
Whole gameplay until this save is visible [here](https://www.youtube.com/playlist?list=PL-Q1t_Y2KQTXLxLcASKisUmeQqll1GKe-), save at the end of part 03 is the one that doesn't load.

Here are the errors displayed by PCSX2 console:
```
	HotSwapping to new ISO src image!
HLE Notice: ELF does not have a path.

Opening plugins...
	Opening GS
	Opening PAD
	Opening SPU2
3.x GL context successfully created
Current Renderer: OpenGL (Hardware renderer)
Available VRAM/RAM:3840MB for textures
GSdx Lookup CRC:9A51B627
GSdx Lookup CRC:9A51B627
	Opening CDVD
isoFile open ok: C:\Users\romar\Downloads\Project Zero\ISOs\Fatal Frame II (U) (Undub v1.0.0)\Fatal Frame II (U) (Undub v1.0.0).iso
	Image type  = DVD
isoFile: Invalid layer1 Primary Volume Descriptor
 * CDVD Disk Open: DVD, Single layer or unknown:
 * * Track 1: Data (Mode 1) (1627211 sectors)
	Opening USB
	Opening FW
	Opening DEV9
McdSlot 0 [File]: C:\Users\romar\OneDrive\Documents\PCSX2\memcards\Mcd001.ps2
McdSlot 1 [File]: C:\Users\romar\OneDrive\Documents\PCSX2\memcards\Mcd002.ps2
Plugins opened successfully.
	Bios Found: USA     v02.30(20/02/2008)  Console
	BIOS rom2 module not found, skipping...
	BIOS erom module not found, skipping...
# Initialize memory (rev:4.00, ctm:393Mhz, cpuclk:295Mhz detected)

PlayStation 2 ======== Hard reset boot
 ROMGEN=2008-0220, IOP info (CPUID=1f, CACH_CONFIG=14000200, 2MB, IOP mode)
 <20080220-175343,ROMconf,PS20230WC20080220.bin:11824>
# Total accessable memory size: 32 MB (B:2:8:0) (400:2:7c30)
# TLB spad=0 kernel=1:12 default=13:30 extended=31:38
# Initialize Start.
# Initialize GS ...
# Initialize INTC ...
# Initialize TIMER ...
# Initialize DMAC ...
# Initialize VU1 ...
# Initialize VIF1 ...
# Initialize GIF ...
# Initialize VU0 ...
# Initialize VIF0 ...
# Initialize IPU ...
# Initialize FPU ...
# Initialize User Memory ...
# Initialize Scratch Pad ...
# Initialize Done.

EE DECI2 Manager version 0.06 Sep  5 2006 12:57:08
  CPUID=2e20, BoardID=0, ROMGEN=2008-0220, 32M

Patches: No CRC found, using 00000000 instead.
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC
found 16260 symbols
ELF (cdrom0:\SLUS_207.66;1) Game CRC = 0x1C6C1D8E, EntryPoint = 0x00100008
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC


IOP Realtime Kernel Ver.0.9.1

    Copyright 1999 (C) Sony Computer Entertainment Inc. 

Reboot service module.(99/11/10)
cdvd driver module version 0.1.1 (C)SCEI
Load File service.(99/11/05)
Multi Threaded Fileio module.(99/11/15) 
iop heap service (99/11/03)
loadelf: fname rom0:OSDSYS secname all
loadelf version 3.30
Input ELF format filename = rom0:OSDSYS
0 00100000 00058530 ......
Loaded, rom0:OSDSYS
start address 0x100008
gp address 00000000
# Restart Without Memory Clear.
# Initialize GS ...
# Initialize INTC ...
# Initialize TIMER ...
# Initialize DMAC ...
# Initialize VU1 ...
# Initialize VIF1 ...
# Initialize GIF ...
# Initialize VU0 ...
# Initialize VIF0 ...
# Initialize IPU ...
# Initialize FPU ...
# Initialize Scratch Pad ...
# Restart Without Memory Clear Done.
# Restart Without Memory Clear.
# Initialize GS ...
# Initialize INTC ...
# Initialize TIMER ...
# Initialize DMAC ...
# Initialize VU1 ...
# Initialize VIF1 ...
# Initialize GIF ...
# Initialize VU0 ...
# Initialize VIF0 ...
# Initialize IPU ...
# Initialize FPU ...
# Initialize Scratch Pad ...
# Restart Without Memory Clear Done.
Get Reboot Request From EE

PlayStation 2 ======== Update rebooting..

PlayStation 2 ======== Update reboot complete
cdvdman Init
rmreset start
rmreset end
clearspu: completed
Pad Driver for OSD (2000/12/05 18:05) 392, 128


IOP Realtime Kernel Ver.0.9.1

    Copyright 1999 (C) Sony Computer Entertainment Inc. 

Reboot service module.(99/11/10)
cdvd driver module version 0.1.1 (C)SCEI
Load File service.(99/11/05)
Multi Threaded Fileio module.(99/11/15) 
iop heap service (99/11/03)
read/write allocate memory 4000
open name rom0:OSDVER flag 1 data 4aa78
open fd = 2
open name rom0:ROMVER flag 1 data 4aa78
open fd = 2
open name rom0:XDEV9 flag 1 data 4aa78
open fd = 2
loadmodule: fname rom0:XDEV9 args 0 arg 
dev9: unknown dev9 hardware.
loadmodule: id 32, ret 1
loadmodule: fname rom0:XDEV9SERV args 0 arg 
loadmodule: id -200, ret 1
open name rom0:ROMVER flag 1 data 4aa78
open fd = 2
open name rom0:FONTM flag 1 data 4aa78
open fd = 2
open name rom0:FNTIMAGE flag 1 data 4aa78
open fd = 2
open name rom0:SNDIMAGE flag 1 data 4aa78
open fd = 2
open name rom0:TEXIMAGE flag 1 data 4aa78
open fd = 2
open name rom0:ICOIMAGE flag 1 data 4aa78
open fd = 2
open name rom0:TZLIST flag 1 data 4aa78
open fd = 2
_spu_rev_offsetaddr 1f9021
### addr = 1034255  size = 28640
_spu_rev_offsetaddr 1d9021
### addr = 968719  size = 28640
get val add 3c0 type 2 ret 1ff000
get val add 1ff000 type 2 ret 300
open name rom0:PS1ID flag 1 data 4aa78
open fd = 2
open name rom0:PS1VERA flag 1 data 4aa78
open fd = 2
open name rom1:DVDIDU flag 1 data 4aa78
open fd = 2
open name rom1:DVDVERU flag 1 data 4aa78
open fd = 2
(UpdateVSyncRate) Mode Changed to DVD NTSC.
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC
open name cdrom0:¥SYSTEM.CNF;1 flag 5 data 4aa78
open fd = 2
ExecutePs2GameDisk
open name cdrom0:¥SYSTEM.CNF;1 flag 5 data 4aa78
open fd = 2
# Restart.
# Initialize GS ...
# Initialize INTC ...
# Initialize TIMER ...
# Initialize DMAC ...
# Initialize VU1 ...
# Initialize VIF1 ...
# Initialize GIF ...
# Initialize VU0 ...
# Initialize VIF0 ...
# Initialize IPU ...
# Initialize FPU ...
# Initialize User Memory ...
# Initialize Scratch Pad ...
# Restart Done.
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC
Get Reboot Request From EE

PlayStation 2 ======== Update rebooting..

PlayStation 2 ======== Update reboot complete
cdvdman Init


IOP Realtime Kernel Ver.0.9.1

    Copyright 1999 (C) Sony Computer Entertainment Inc. 

Reboot service module.(99/11/10)
cdvd driver module version 0.1.1 (C)SCEI
Load File service.(99/11/05)
Multi Threaded Fileio module.(99/11/15) 
iop heap service (99/11/03)
loadelf: fname rom0:PS2LOGO secname all
loadelf version 3.30
Input ELF format filename = rom0:PS2LOGO
0 01000000 000172c0 ..
Loaded, rom0:PS2LOGO
start address 0x1000008
gp address 00000000
# Restart Without Memory Clear.
# Initialize GS ...
# Initialize INTC ...
# Initialize TIMER ...
# Initialize DMAC ...
# Initialize VU1 ...
# Initialize VIF1 ...
# Initialize GIF ...
# Initialize VU0 ...
# Initialize VIF0 ...
# Initialize IPU ...
# Initialize FPU ...
# Initialize Scratch Pad ...
# Restart Without Memory Clear Done.
read/write allocate memory 4000
open name rom0:ROMVER flag 1 data 45d78
open fd = 2
loadmodule: fname rom0:OSDSND args 0 arg 
loadmodule: id 25, ret 0
open name cdrom0:¥SYSTEM.CNF;1 flag 5 data 45d78
open fd = 2
loadmodule: fname rom0:XPARAM args 12 arg SLUS_207.66
loadmodule: id 26, ret 1
_spu_rev_offsetaddr 1dfff1
### addr = 2097120  size = 32
_spu_rev_offsetaddr 1f9021
### addr = 1034255  size = 28640
(UpdateVSyncRate) Mode Changed to NTSC.
# Restart.
# Initialize GS ...
# Initialize INTC ...
# Initialize TIMER ...
# Initialize DMAC ...
# Initialize VU1 ...
# Initialize VIF1 ...
# Initialize GIF ...
# Initialize VU0 ...
# Initialize VIF0 ...
# Initialize IPU ...
# Initialize FPU ...
# Initialize User Memory ...
# Initialize Scratch Pad ...
# Restart Done.
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC
Get Reboot Request From EE

PlayStation 2 ======== Update rebooting..

PlayStation 2 ======== Update reboot complete
cdvdman Init


IOP Realtime Kernel Ver.0.9.1

    Copyright 1999 (C) Sony Computer Entertainment Inc. 

Reboot service module.(99/11/10)
cdvd driver module version 0.1.1 (C)SCEI
Load File service.(99/11/05)
Multi Threaded Fileio module.(99/11/15) 
iop heap service (99/11/03)
loadelf: fname cdrom0:¥SLUS_207.66;1 secname all
loadelf version 3.30
(SYSTEM.CNF) Detected PS2 Disc = cdrom0:\SLUS_207.66;1
(SYSTEM.CNF) Software version = 1.00
(SYSTEM.CNF) Disc region type = NTSC
Input ELF format filename = cdrom0:¥SLUS_207.66;1
0 00100000 0016b39c .......................
1 0026b400 000d8b90 ...............
Loaded, cdrom0:¥SLUS_207.66;1
start address 0x100008
gp address 00000000
# Restart Without Memory Clear.
# Initialize GS ...
# Initialize INTC ...
# Initialize TIMER ...
# Initialize DMAC ...
# Initialize VU1 ...
# Initialize VIF1 ...
# Initialize GIF ...
# Initialize VU0 ...
# Initialize VIF0 ...
# Initialize IPU ...
# Initialize FPU ...
# Initialize Scratch Pad ...
# Restart Without Memory Clear Done.
Elf entry point @ 0x00100008 about to get recompiled. Load patches first.
Get Reboot Request From EE
GSdx Lookup CRC:1C6C1D8E
GSdx Lookup CRC:1C6C1D8E

PlayStation 2 ======== Update rebooting..

PlayStation 2 ======== Update reboot complete
cdvdman Init


IOP Realtime Kernel Ver. 2.2

    Copyright 1999-2002 (C) Sony Computer Entertainment Inc. 

Reboot service module.(99/11/10)
cdvd driver module version 0.1.1 (C)SCEI
Load File service.(99/11/05)
Multi Threaded Fileio module.(99/11/15) 
iop heap service (99/11/03)
SDR driver version 4.0.1 (C) SCEI
 Exit rsd_main 
(EE pc:00100338) TLB Miss, addr=0x0 [store]
(EE pc:00100338) TLB Miss, addr=0x24 [store]
(EE pc:00100338) TLB Miss, addr=0x48 [store]
(EE pc:00100338) TLB Miss, addr=0x6c [store]
(EE pc:00100338) TLB Miss, addr=0x90 [store]
(EE pc:00100338) TLB Miss, addr=0xb4 [store]
(EE pc:00100338) TLB Miss, addr=0xd8 [store]
(EE pc:00100338) TLB Miss, addr=0xfc [store]
(EE pc:00100338) TLB Miss, addr=0x120 [store]
(EE pc:00100338) TLB Miss, addr=0x144 [store]
(EE pc:00100338) TLB Miss, addr=0x168 [store]
(EE pc:00100338) TLB Miss, addr=0x18c [store]
(EE pc:00100338) TLB Miss, addr=0x1b0 [store]
(EE pc:00100338) TLB Miss, addr=0x1d4 [store]
(EE pc:00100338) TLB Miss, addr=0x1f8 [store]
(EE pc:00100338) TLB Miss, addr=0x21c [store]
(EE pc:00100338) TLB Miss, addr=0x240 [store]
(EE pc:00100338) TLB Miss, addr=0x264 [store]
(EE pc:00100338) TLB Miss, addr=0x288 [store]
(EE pc:00100338) TLB Miss, addr=0x2ac [store]
(EE pc:00100338) TLB Miss, addr=0x2d0 [store]
(EE pc:00100338) TLB Miss, addr=0x2f4 [store]
(EE pc:00100338) TLB Miss, addr=0x318 [store]
(EE pc:00100338) TLB Miss, addr=0x33c [store]
(EE pc:00100338) TLB Miss, addr=0x360 [store]
(EE pc:00100338) TLB Miss, addr=0x384 [store]
(EE pc:00100338) TLB Miss, addr=0x3a8 [store]
(EE pc:00100338) TLB Miss, addr=0x3cc [store]
(EE pc:00100338) TLB Miss, addr=0x3f0 [store]
(EE pc:00100338) TLB Miss, addr=0x414 [store]
(EE pc:00100338) TLB Miss, addr=0x438 [store]
(EE pc:00100338) TLB Miss, addr=0x45c [store]
(EE pc:00100338) TLB Miss, addr=0x480 [store]
(EE pc:00100338) TLB Miss, addr=0x4a4 [store]
(EE pc:00100338) TLB Miss, addr=0x4c8 [store]
(EE pc:00100338) TLB Miss, addr=0x4ec [store]
(EE pc:00100338) TLB Miss, addr=0x510 [store]
(EE pc:00100338) TLB Miss, addr=0x534 [store]
(EE pc:00100338) TLB Miss, addr=0x558 [store]
(EE pc:00100338) TLB Miss, addr=0x57c [store]
(EE pc:00100338) TLB Miss, addr=0x5a0 [store]
(EE pc:00100338) TLB Miss, addr=0x5c4 [store]
(EE pc:00100338) TLB Miss, addr=0x5e8 [store]
(EE pc:00100338) TLB Miss, addr=0x60c [store]
(EE pc:00100338) TLB Miss, addr=0x630 [store]
(EE pc:00100338) TLB Miss, addr=0x654 [store]
(EE pc:00100338) TLB Miss, addr=0x678 [store]
(EE pc:00100338) TLB Miss, addr=0x69c [store]
(EE pc:00100338) TLB Miss, addr=0x6c0 [store]
(EE pc:00100338) TLB Miss, addr=0x6e4 [store]
Closing plugins...
	Closing DEV9
	Closing FW
	Closing USB
	Closing CDVD
	Closing SPU2
	Closing PAD
	Closing GS
Delete 0 Shaders, 78 Programs, 28 Pipelines
Plugins closed successfully.
Decommitting host memory for virtual systems...
EE/iR5900-32 Recompiler Reset
```

**Comment by wagrenier** [comment]
After investigation, the issue seems to come from a memory allocation failed from a function called `SAFE_MALLOC` which is supposed to return a memory address which then gets zeroed out. However, with the undub, something seems to go wrong at that allocation and returns an address of `0x00000000` which makes the game crash on real hardware but not on PCSX2.

This could be due to too much data being loaded at once with some larger file size in the Japanese version. What file causes this issue should be investigated and made to not be undubbed for a hardware release of the undub.

**Comment by wagrenier** [comment]
Fixed in (6edd467779)[https://github.com/wagrenier/Zero2Undub/commit/6edd467982e55a31f70a87b658c2c5564d1ddf38]
